### ISSUE TITLE
Roslyn server command line options

### ISSUE THREAD

**Issue by leblocks** [seblj/roslyn.nvim#92]
Hello there, 

I have a pretty complicated solution with lots of csprojs and binaries all over the place, roslyn fails to find them all and I have a feeling that I may fix it with some command line parameters passed to roslyn, unfortunately I can't find any proper documentation on it.

Do you know where can I get one?

**Comment by seblj** [comment]
No not really...

Do you have any concrete example, and then I can look to see if there is something I am doing wrong, or can extend do support or something?

**Comment by leblocks** [comment]
I'll try to post tomorrow some logs from roslyn, maybe it will shed some light. Sorry for an obscure description of the issue.

**Comment by leblocks** [comment]
I have a very large solutions (~400 csproj) all other lsp's failed me :D (I've been using omnisharp for 2 years and after that csharp_ls for a little bit)

Note: those projects are using .NET 4.7.6 framework.
After I launch nvim and roslyn server starts, it takes some time initialize and it finishes. Basic navigation (in the same project works) but not more than that. It behaves as if solution have not been built.

lsp log has lots of errors like that:

```text
[ERROR][2024-11-26 12:50:14] ...lsp/handlers.lua:623	'[LanguageServerProjectSystem] Error while loading //path/to/some/project.csproj: The task factory "SOMETASKNAME" is not supported on the .NET Core version of MSBuild.'
[ERROR][2024-11-26 12:50:14] ...lsp/handlers.lua:623	'[LanguageServerProjectSystem] Error while loading //path/to/some/project.csproj: The "SOMETASKNAME" task was not found. Check the following: 1.) The name of the task in the project file is the same as the name of the task class. 2.) The task class is "public" and implements the Microsoft.Build.Framework.ITask interface. 3.) The task is correctly declared with <UsingTask> in the project file, or in the *.tasks files located in the "C:\\Program Files\\dotnet\\sdk\\8.0.404" directory.'
[ERROR][2024-11-26 12:50:14] ...lsp/handlers.lua:623	'[LanguageServerProjectSystem] Error while loading //path/to/some/project.csproj: The task factory "SOMETASKNAME" is not supported on the .NET Core version of MSBuild.'
```

So I have a feeling that it is some configuration issue.

**Comment by seblj** [comment]
Hmm yeah okay I have no idea then... I also have a large project with a lot of csproj files with .NET Framework 4.8.1, but I have given up on using neovim with that old project on windows. I have never really been able to comfortably use neovim om windows, and also never got omnisharp and other language servers working for legacy .NET Framework...

**Comment by leblocks** [comment]
Ok, got you. I'm prety comfortable with neovim on windows it works fine. BTW, core projects work just fine on windows with roslyn. Thanks for the responses.

P.S. does roslyn.nvim has an option to disable autostart?

**Comment by seblj** [comment]
No problem!

> P.S. does roslyn.nvim has an option to disable autostart?

No, maybe it is possible to do, but no option. However, I am open to adding it. How do you think it should work?

I am thinking of maybe `:Roslyn start` command then. But then, after starting, I guess it should autostart every buffer before doing `:Roslyn stop`.

Currently if you are doing `:Roslyn stop` it will attach to the next buffer again, but if autostart is disabled, then it should probably not start until doing `:Roslyn start` again?

**Comment by drzbida** [comment]
Is the plugin slow on large solutions, or the language server itself? I'm working on a ~40 csproj solution so I didn't hit any performance issues yet, just a few server crashes here and there which I'm starting to try debug.

I was thinking that if the plugin itself is slow, the approach of finding the appropriate csproj and sln (that is done even with solution locked / selected) could be changed from being done on every file entry to being done once on plugin init. 

We could build one or two trees for the csproj and sln files with nodes that are either folders or our csproj/sln files. Then the best match for the current buffer would be found in O(number of path components) of the current buffer relative to the root dir. These could be built very fast if we add a dependency requirement to fd, which most people have installed since it's required in a lot of popular plugins. If that's not acceptable, I think waiting once a bit longer is still better than always waiting long. We could even cache these persistently on the disk. The only a bit harder parts I could see is finding out when out cache is no longer valid due to new / removed csprojs/slns. We could probably hash the slns to quickly figure out if something changed. There are probably more edge cases I can't think of right now but I don't think there's anything unsolvable. 

I could try implementing this sometime next month if it is of interest, after I figure out why my server crashes and what could be done / after I give up :) 

In the meantime, I think an easy performance improvement would be to first check in the FileType autocmd if the target is locked or the solution is selected and return fast, and do the utils.root_dir search directly in the target subcommand just before the vim.ui.select, if there isn't any other reason why it's not done like this

**Comment by leblocks** [comment]
From my point of view, if _autostart_ is disabled do not start anything. Once you started roslyn manually - regular behaviour as it works today.

**Comment by seblj** [comment]
> From my point of view, if autostart is disabled do not start anything. Once you started roslyn manually - regular behaviour as it works today.

Okay, that seems fair. I can look into implementing this soon :)

**Comment by drzbida** [comment]
>
> Then I think I will try to plug in a small async library here to resolve the targets async

I think https://github.com/nvim-neotest/nvim-nio is starting to become the standard for this. Neotree has a refactor branch which uses it, which means most users will already have it installed. I think it could also be used to simplify the codebase from all the callbacks https://github.com/pysan3/neo-tree.nvim/blob/v4-dev/MIGRATION-v4.md#nio-async-vs-callbacks

**Comment by seblj** [comment]
Yeah, I think I will go down the route of just bundling a small async library inside here like I have done here:

https://github.com/seblj/nvim-formatter/blob/master/lua/formatter/async.lua

**Comment by seblj** [comment]
So I fixed it so it does an early return now. I have no idea why I didn't do that from the start.

Also, I wrappes the logic for finding root and the following code in a vim.schedule which seems to help with the perception of slowness at least. It seems like it will schedule this until after the UI is initialized which makes it more "instant" in loading a buffer.

There is probably more that can be done, and actually implementing it asynchronously. I am however not sure how much effort I would put into this myself because it would take me more effort than I think it is worth.

If it's still anoyingly slow, I would accept a PR to fix it ofc. If I where to choose, I think I would like to depend on `fd` over a third party async library. Of course, no dependencies would be even better if possible.

As for the autostart, I got busy and didn't have time to do it now in a clean way. I am having this in the back of my mind, thinking about how I want to implement this, and hopefully I can do it soon

**Comment by drzbida** [comment]
>
> As for the autostart, I got busy and didn't have time to do it now in a clean way. I am having this in the back of my mind, thinking about how I want to implement this, and hopefully I can do it soon

Technically most of the work is already done by lazy.nvim. With a user setup like this it works exactly as described

![image](https://github.com/user-attachments/assets/e01fc657-f870-4ce5-83e0-d456dabb690c)

The only bad part about this is that the user needs to have internal plugin knowledge to exec the correct autocmds. If the roslyn commands were created in the setup instead of the FileType autocmd, he wouldn't need to know anything but the API, he would just call a :Roslyn start, or the :Roslyn restart could be changed to start the server if it's there's no client connected (I think that's the same behavior as LspRestart, but I might be wrong ? )

![image](https://github.com/user-attachments/assets/a624bbaa-1d2f-4a0a-8873-9cd92efe729f)

**Comment by seblj** [comment]
But wouldn't the problem be that the plugin still need to be loaded for the commands to be available?

That means that the autocmd would be created and then attach to the buffer without any type of guard from the config?

**Comment by drzbida** [comment]
I'm not sure what you mean, this is what i'm talking about

https://github.com/seblj/roslyn.nvim/compare/main...drzbida:roslyn.nvim:init-commands-on-setup

And then the user would do something like this if he/she wants to not start automatically

![image](https://github.com/user-attachments/assets/89c9e3ec-9887-4097-a4d4-0954a0a49c07)

I don't quite understand exactly why the commands need to be attached to the cs buffer, usually when you use a plugin you expect to have the same API available all the time. The same goes for the target subcommand, which I didn't change as it was needed for the example. If you do a target while inside a csproj for example, why shouldn't it work ? This one should probably only check if the buffer is indeed related to a file and just return otherwise

I think that all the autocmd should do is attach the lsp to the buffer / choose the correct sln as that's the closest thing to what a normal lsp does; the commands should be unrelated imo, but I might be wrong

**Comment by seblj** [comment]
> I don't quite understand exactly why the commands need to be attached to the cs buffer

I agree. I did some mistakes earlier when I initially introdced the `:CSTarget` command, but now I think it is in a good place to easily refactor it out. I'll look into it👍

> And then the user would do something like this if he/she wants to not start automatically

Yeah, but then I think that the autostart would be very coupled with lazy.nvim. I ideally then would want something like this to work

```lua
{
    "seblj/roslyn.nvim,
    keys = { "<leader>ft" }
}
```
